SmartDeviceLink/SDLMenuManager.m

```diff
@@ -0,0 +1,416 @@
+    SDLAddCommand *command = [[SDLAddCommand alloc] init];
+
+    SDLMenuParams *params = [[SDLMenuParams alloc] init];
+    params.menuName = cell.title;
+    params.parentID = cell.parentCellId != UINT32_MAX ? @(cell.parentCellId) : nil;
+    params.position = @(position);
+
+    command.menuParams = params;
+    command.vrCommands = cell.voiceCommands;
+    command.cmdIcon = (cell.icon && shouldHaveArtwork) ? [[SDLImage alloc] initWithName:cell.icon.name] : nil;
+    command.cmdID = @(cell.cellId);
+
+    return command;
+}
+
+- (SDLAddSubMenu *)sdl_subMenuCommandForMenuCell:(SDLMenuCell *)cell position:(UInt16)position {
+    SDLAddSubMenu *submenu = [[SDLAddSubMenu alloc] initWithId:cell.cellId menuName:cell.title];
+    submenu.position = @(position);
+
+    return submenu;
+}
+
+#pragma mark - Observers
+
+- (void)sdl_commandNotification:(SDLRPCNotificationNotification *)notification {
+    SDLOnCommand *onCommand = (SDLOnCommand *)notification.notification;
+
+    for (SDLMenuCell *cell in self.menuCells) {
+        if (onCommand.cmdID.unsignedIntegerValue != cell.cellId) { continue; }
+
+        cell.handler(onCommand.triggerSource);
+        break;
+    }
+}
+
+- (void)sdl_registerResponse:(SDLRPCResponseNotification *)notification {
+    SDLRegisterAppInterfaceResponse *response = (SDLRegisterAppInterfaceResponse *)notification.response;
+    self.displayCapabilities = response.displayCapabilities;
+}
+
+- (void)sdl_displayLayoutResponse:(SDLRPCResponseNotification *)notification {
+    SDLSetDisplayLayoutResponse *response = (SDLSetDisplayLayoutResponse *)notification.response;
+    self.displayCapabilities = response.displayCapabilities;
+}
+
+- (void)sdl_hmiStatusNotification:(SDLRPCNotificationNotification *)notification {
+    SDLOnHMIStatus *hmiStatus = (SDLOnHMIStatus *)notification.notification;
+
+    SDLHMILevel oldHMILevel = self.currentHMILevel;
+    self.currentHMILevel = hmiStatus.hmiLevel;
+
+    // Auto-send an updated menu if we were in NONE and now we are not, and we need an update
+    if ([oldHMILevel isEqualToString:SDLHMILevelNone] && ![self.currentHMILevel isEqualToString:SDLHMILevelNone]) {
+        if (self.waitingOnHMIUpdate) {
+            [self setMenuCells:_menuCells];
+        } else {
+            [self sdl_updateWithCompletionHandler:nil];
+        }
+    }
+
+    // If we don't check for this and only update when not in the menu, there can be IN_USE errors, especially with submenus. We also don't want to encourage changing out the menu while the user is using it for usability reasons.
+    SDLSystemContext oldSystemContext = self.currentSystemContext;
+    self.currentSystemContext = hmiStatus.systemContext;
+
+    if ([oldSystemContext isEqualToEnum:SDLSystemContextMenu] && ![self.currentSystemContext isEqualToEnum:SDLSystemContextMenu]) {
+        if (self.waitingOnHMIUpdate) {
```

Comment:
#### Test cases
I have two separate menus. 
1. Menu 1 has one `AddCommand`
1. Menu 2 is empty 

#### Test Flow
1. Send menu 1
1. Open the menu
1. Send menu 2 (menu 1 should still be open)
1. Close the menu

#### Bug Description
The menu is never deleted. This is because when `[self setMenuCells:_menuCells]` is called from line 409, `oldMenuCells` (line 231) gets set to `@[]` before the function to delete the old `AddCommand`s is called.